### PR TITLE
pythonPackages.pysnooper: init at 0.3.0

### PR DESCRIPTION
**Pull request by seqizz**
###### Motivation for this change

pysnooper is a python logging library, providing quick logging via decorator

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/pysnooper/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  # Checks require python_toolbox, which has failing its own checks
+  # Issue reported upstream
+  doCheck = false;
+  propagatedBuildInputs = [
+    six
+    future
+    decorator
+  ];
+
+  meta = with stdenv.lib; {
```
Comment (dotlambda) [inline pkgs/development/python-modules/pysnooper/default.nix]:
```suggestion
  meta = with lib; {
```

File: pkgs/development/python-modules/pysnooper/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+, buildPythonPackage
+, fetchPypi
+, six
+, future
+, decorator
+}:
+
+buildPythonPackage rec {
+  version = "0.0.21";
+  pname = "PySnooper";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pysnooper/default.nix]:
```suggestion
  pname = "pysnooper";
```
You'll have to specify it again within `fetchPypi`.

File: pkgs/development/python-modules/pysnooper/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "15vyc6hpy5whsswmz3a5qrcrcgd59lxf4q56krkcdzmk88agnwpg";
+  };
+
+  # Checks require python_toolbox, which has failing its own checks
+  # Issue reported upstream
```
Comment (dotlambda) [inline pkgs/development/python-modules/pysnooper/default.nix]: Please package python_toolbox and deselect the failing tests.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/pysnooper/default.nix (hunk 1 above)
Comment (seqizz) [reply pkgs/development/python-modules/pysnooper/default.nix]: I currently pushed this change. Sorry for my ignorance but when I change this and try to build with:
```
nix-build -I ~/devel/nixpkgs -E "with import <nixpkgs> {}; python3Packages.callPackage ~/devel/nixpkgs/pkgs/development/python-modules/pysnooper/default.nix { }"
```
I get: "error: undefined variable 'lib' at ..."
Am I missing something?

Comment (dotlambda) [reply pkgs/development/python-modules/pysnooper/default.nix]: Yes, you'll have to replace `stdenv` by `lib` in the arguments.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please make a separate commit for each package.

File: pkgs/development/python-modules/python_toolbox/default.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    rm source_py{2,3}/test_python_toolbox/test_combi/test_extensive.py
+    rm source_py{2,3}/test_python_toolbox/test_cute_testing/test_raise_assertor.py
+  '';
+
+  propagatedBuildInputs = [
+    docutils
+    nose
+  ];
+
+  meta = with stdenv.lib; {
```
Comment (dotlambda) [inline pkgs/development/python-modules/python_toolbox/default.nix]:
```suggestion
  meta = with lib; {
```

File: pkgs/development/python-modules/python_toolbox/default.nix [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    description = "A collection of Python tools for various tasks";
```
Comment (dotlambda) [inline pkgs/development/python-modules/python_toolbox/default.nix]: Please be more descriptive.

File: pkgs/development/python-modules/pysnooper/default.nix [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    sha256 = "15vyc6hpy5whsswmz3a5qrcrcgd59lxf4q56krkcdzmk88agnwpg";
+  };
+
+  propagatedBuildInputs = [
+    six
+    future
+    decorator
+  ];
+
+  buildInputs = [
```
Comment (dotlambda) [inline pkgs/development/python-modules/pysnooper/default.nix]: `checkInputs`?

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/pysnooper/default.nix (hunk 1 above)
Comment (seqizz) [reply pkgs/development/python-modules/pysnooper/default.nix]: Ah.. now the python_toolbox is required argument, but should be ok now. Thanks.

File: pkgs/development/python-modules/python_toolbox/default.nix (hunk 5 above)
Comment (seqizz) [reply pkgs/development/python-modules/python_toolbox/default.nix]: I guess it's good enough now, considering even the project doesn't have something satisfactory.

File: pkgs/development/python-modules/pysnooper/default.nix (hunk 6 above)
Comment (seqizz) [reply pkgs/development/python-modules/pysnooper/default.nix]: Right, sorry, rookieness.. :)

**Comment by seqizz** [comment]
Should be ok know, could you re-check. Also thanks for patience.

File: pkgs/top-level/python-packages.nix [hunk 7]
```diff
@@ -5388,6 +5388,10 @@ in {
 
   pysensors = callPackage ../development/python-modules/pysensors { };
 
+  python_toolbox = callPackage ../development/python-modules/python_toolbox { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]:
```suggestion
  python-toolbox = callPackage ../development/python-modules/python-toolbox { };
```
See https://nixos.org/nixpkgs/manual/#contributing.

File: pkgs/development/python-modules/python_toolbox/default.nix [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    rm source_py{2,3}/test_python_toolbox/test_cute_testing/test_raise_assertor.py
+  '';
+
+  propagatedBuildInputs = [
+    docutils
+    nose
+  ];
+
+  meta = with lib; {
+    description = "100+ helper tools for Python development";
```
Comment (dotlambda) [inline pkgs/development/python-modules/python_toolbox/default.nix]: How about
```suggestion
    description = "Tools for testing PySnooper";
```

File: pkgs/development/python-modules/python_toolbox/default.nix [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  };
+
+  # https://github.com/cool-RR/python_toolbox/issues/3
+  # These tests are failing
+  preCheck = ''
+    rm source_py{2,3}/test_python_toolbox/test_combi/test_extensive.py
+    rm source_py{2,3}/test_python_toolbox/test_cute_testing/test_raise_assertor.py
+  '';
+
+  propagatedBuildInputs = [
```
Comment (dotlambda) [inline pkgs/development/python-modules/python_toolbox/default.nix]:
```suggestion
  checkInputs = [
```

File: pkgs/development/python-modules/pysnooper/default.nix [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+, buildPythonPackage
+, fetchPypi
+, six
+, future
+, decorator
+, python_toolbox
+}:
+
+buildPythonPackage rec {
+  version = "0.0.21";
```
Comment (dotlambda) [inline pkgs/development/python-modules/pysnooper/default.nix]: 0.0.22 was released

File: pkgs/development/python-modules/pysnooper/default.nix [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  src = fetchPypi {
+    inherit version;
+    pname = "PySnooper";
+    sha256 = "15vyc6hpy5whsswmz3a5qrcrcgd59lxf4q56krkcdzmk88agnwpg";
+  };
+
+  propagatedBuildInputs = [
+    six
+    future
+    decorator
```
Comment (dotlambda) [inline pkgs/development/python-modules/pysnooper/default.nix]: The new version claims to have no dependencies.

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by seqizz** [comment]
Re-pushed with suggested changes

**Review by dotlambda — CHANGES_REQUESTED** [review]
The first commit message needs to be amended.

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.pysnooper python3.pkgs.pysnooper

File: pkgs/development/python-modules/python-toolbox/default.nix [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  };
+
+  # https://github.com/cool-RR/python_toolbox/issues/3
+  # These tests are failing
+  preCheck = ''
+    rm source_py{2,3}/test_python_toolbox/test_combi/test_extensive.py
+    rm source_py{2,3}/test_python_toolbox/test_cute_testing/test_raise_assertor.py
+  '';
+
+  checkInput = [
```
Comment (dotlambda) [inline pkgs/development/python-modules/python-toolbox/default.nix]:
```suggestion
  checkInputs = [
```

**Review by dotlambda — CHANGES_REQUESTED** [review]

**Comment by seqizz** [comment]
Pushed again

**Comment by dotlambda** [comment]
@seqizz https://github.com/cool-RR/python_toolbox/issues/3 seems to have been resolved.

**Comment by seqizz** [comment]
Switched to `fetchFromGitHub` to get the latest release. Yet not sure how to test this :confused: I can't see the tests on pysnooper build when I try, since there is no way to say "force rebuild" on nix-shell or nix-build as far as I can find.

**Comment by seqizz** [comment]
Pushed 0.1.0

**Comment by FRidh** [comment]
@grahamcofborg build pythonPackages.pysnooper python3Packages.pysnooper

**Comment by seqizz** [comment]
I think github is confused, suggested change is already done..
I wish I could've helped maintaining the nixos repo more, since I can see the MR/issue pile :no_mouth: But you can see the quality of my commits already :grimacing:

**Comment by seqizz** [comment]
After this much time, I guess I know a bit more what I am doing :smiley: 
Upgraded to latest version, also built fine on my box.

File: pkgs/development/python-modules/python-toolbox/default.nix [hunk 13]
```diff
@@ -0,0 +1,30 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, nose
+, docutils
+}:
+
+buildPythonPackage rec {
+  version = "0.9.4";
+  pname = "python_toolbox";
```
Comment (jonringer) [inline pkgs/development/python-modules/python-toolbox/default.nix]:
```suggestion
  pname = "python_toolbox";
  disabled = isPy27;
```

looks like they don't support it:
```
builder for '/nix/store/cqj58n5ihgxkkmmk0qxr9rmjnqwph0z4-python2.7-python_toolbox-0.9.4.drv' failed with exit code 1; last 10 log lines:
    File "/build/source/source_py2/test_python_toolbox/test_future_tools/test_future_tools.py", line 4, in <module>
      import concurrent.futures
  ImportError: No module named concurrent.futures

  ----------------------------------------------------------------------
  Ran 522 tests in 199.426s

  FAILED (errors=1, skipped=7)
  Test failed: <unittest.runner.TextTestResult run=522 errors=1 failures=0>
  error: Test failed: <unittest.runner.TextTestResult run=522 errors=1 failures=0>
cannot build derivation '/nix/store/pzk44l92xshsmzasq76vd6y1c4c72jqa-python2.7-pysnooper-0.3.0.drv': 1 dependencies couldn't be built
```

Comment (seqizz) [reply pkgs/development/python-modules/python-toolbox/default.nix]: Ah, thanks. I've also disabled tests on pysnooper with same condition :+1:

**Review by jonringer — APPROVED** [review]
diff LGTM
commits LGTM
has tests :+1: 

failures appear to be upstream issues
```
[5 built (1 failed), 0.0 MiB DL]
error: build of '/nix/store/c07fmi7g9j9ggfy9jkwcddq53ycny9g0-env.drv' failed
https://github.com/NixOS/nixpkgs/pull/60228
2 package failed to build:
python38Packages.pysnooper python38Packages.python-toolbox

3 package built:
python27Packages.pysnooper python37Packages.pysnooper python37Packages.python-toolbox
```

**Comment by jonringer** [comment]
@GrahamcOfBorg build python27Packages.pysnooper python37Packages.pysnooper python37Packages.python-toolbox

**Pull request merged**